pallets/identity/src/claims.rs

```diff
@@ -99,85 +181,64 @@ impl<T: Config> Module<T> {
     /// No state change is allowed in this function because this function is used within the RPC
     /// calls.
     pub fn fetch_cdd(claim_for: IdentityId, leeway: T::Moment) -> Option<IdentityId> {
-        Self::base_fetch_cdd(claim_for, leeway, None)
+        Self::base_fetch_cdd(claim_for, leeway, None, true)
     }
 
     fn base_fetch_cdd(
         claim_for: IdentityId,
         leeway: T::Moment,
         filter_cdd_id: Option<CddId>,
+        include_parent: bool,
     ) -> Option<IdentityId> {
-        Self::base_fetch_valid_cdd_claims(claim_for, leeway, filter_cdd_id)
+        Self::base_fetch_valid_cdd_claims(claim_for, leeway, filter_cdd_id, include_parent)
             .map(|id_claim| id_claim.claim_issuer)
             .next()
     }
 
-    pub fn base_fetch_valid_cdd_claims(
-        claim_for: IdentityId,
-        leeway: T::Moment,
-        filter_cdd_id: Option<CddId>,
+    // Returns a lazy iterator that will return the CDD claims from the
+    // parent of `did` if they are a child identity.
+    //
+    // If `include_parent` is `false` then the iterator will not return claims
+    // from the parent.
+    pub fn base_fetch_parent_cdd_claims(
+        did: IdentityId,
+        include_parent: bool,
     ) -> impl Iterator<Item = IdentityClaim> {
-        let exp_with_leeway = <pallet_timestamp::Pallet<T>>::get()
-            .checked_add(&leeway)
-            .unwrap_or_default();
-
-        // Supressing `mut` warning since we need mut in `runtime-benchmarks` feature but not otherwise.
-        #[allow(unused_mut)]
-        let mut active_cdds_temp = T::CddServiceProviders::get_active_members();
-
-        // For the benchmarks, self cdd claims are allowed and hence the claim target is added to the cdd providers list.
-        #[cfg(feature = "runtime-benchmarks")]
-        active_cdds_temp.push(claim_for);
-
-        let active_cdds = active_cdds_temp;
-        let inactive_not_expired_cdds = T::CddServiceProviders::get_inactive_members()
-            .into_iter()
-            .filter(|cdd| !T::CddServiceProviders::is_member_expired(cdd, exp_with_leeway))
-            .collect::<Vec<_>>();
-
-        Self::fetch_base_claims(claim_for, ClaimType::CustomerDueDiligence).filter(
-            move |id_claim| {
-                if let Some(cdd_id) = &filter_cdd_id {
-                    if let Claim::CustomerDueDiligence(claim_cdd_id) = &id_claim.claim {
-                        if claim_cdd_id != cdd_id {
-                            return false;
-                        }
-                    }
-                }
+        let mut first_call = include_parent;
+        let mut parent_claims = None;
+        core::iter::from_fn(move || -> Option<IdentityClaim> {
```

Comment:
You can ignore this; I was just thinking about how we can make this a bit simpler and came up with this: 
```Rust
if !include_parent {
    return Vec::new().into_iter();
}

ParentDid::get(did).map_or(Vec::new().into_iter(), |parent_did| {
    let claims: Vec<IdentityClaim> =
        Self::fetch_base_claims(parent_did, ClaimType::CustomerDueDiligence).collect();
    claims.into_iter()
})
```

Reply:
That would cause a read of `ParentDid` and collecting of all the parent claims in the common case (has its own CDD claim).

The `if !include_parent` is an uncommon case, used for InvestorUniqueness CDDId validation (Child identity can't use their parent's CDD for those).

We want most of the work in this function to only happen if the current identity has no valid CDD claims.

Also the most common case only need to get the first valid CDD claim, which is why these functions return iterators.